dispatcher/MongoConnect.py

```diff
@@ -628,9 +634,14 @@ def detector_ackd_command(self, detector, command):
         if (doc := self.collections['outgoing_commands'].find_one(q, sort=sort)) is None:
             self.log.error('No previous command found?')
             return True
-        for h in doc['host']:
-            # loop over doc['host'] because the 'acknowledged' field sometimes
-            # contains extra entries (such as the GPS trigger)
+        # we can't naively use everything in the hosts field, because we might be transitioning
+        # out of linked mode, and there might be "garbage" in the host list because someone
+        # didn't follow very clear instructions, and if a stop is issued to a host that doesn't
+        # exist, the dispatcher basically stops working
+        hosts_this_detector = set(list(self.latest_status[detector]['readers'].keys()) + list(self.latest_status[detector]['controller'].keys()))
+        hosts_in_doc = set(doc['host'])
+        # so we only loop over the intersection of this detector's hosts and the doc's hosts
+        for h in hosts_this_detector & hosts_in_doc:
```

Comment:
Just to be sure, this is one of the key fixes right? So we are going to filter out hosts that are not in the intercept of the sets for the latest status for this logical detector and the command queue for this logical detector.

You think it's also worth logging that we ignore hosts in this hosts-set:
```python
hosts_ignored = hosts_this_detector  ^ hosts_in_doc
if len(hosts_ignored):
   run_id = self.latest_status[detector]['number']
   self.log.warning(f'ignoring {hosts_ignored} for run {run_id}')
```

Reply:
More logging never hurts